File: 0988/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def smallestFromLeaf(self, root: Optional[TreeNode]) -> str:
        def traverse(node, curr):
            val = chr(ord("a") + node.val)
            if not node.left and not node.right:
                return val + curr
            if not node.left:
                return traverse(node.right, val + curr)
            if not node.right:
                return traverse(node.left, val + curr)
            return min(
                traverse(node.left, val + curr), traverse(node.right, val + curr)
            )

        return traverse(root, "")

File: 0988/solution.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

function smallestFromLeaf(root: TreeNode | null): string {
  return traverse(root!, "");
}

function traverse(node: TreeNode, curr: string): string {
  const newWord: string =
    String.fromCharCode("a".charCodeAt(0) + node.val) + curr;
  if (!node.left && !node.right) {
    return newWord;
  }

  if (!node.left) {
    return traverse(node.right!, newWord);
  }

  if (!node.right) {
    return traverse(node.left, newWord);
  }

  const left = traverse(node.left, newWord);
  const right = traverse(node.right, newWord);
  return left < right ? left : right;
}
